pkg/gateway/meta/usecase/vald.go

```diff
@@ -0,0 +1,211 @@
+	eg := errgroup.Get()
+
+	var metadata service.Meta
+
+	metadataClientOptions := append(
+		cfg.Meta.Client.Opts(),
+		grpc.WithErrGroup(eg),
+	)
+
+	var obs observability.Observability
+	if cfg.Observability.Enabled {
+		obs, err = observability.NewWithConfig(cfg.Observability)
+		if err != nil {
+			return nil, err
+		}
+		metadataClientOptions = append(
+			metadataClientOptions,
+			grpc.WithDialOptions(
+				grpc.WithStatsHandler(metric.NewClientHandler()),
+			),
+		)
+	}
+
+	if addrs := cfg.Meta.Client.Addrs; len(addrs) == 0 {
+		return nil, errors.ErrInvalidMetaDataConfig
+	}
+	metadata, err = service.New(
+		service.WithMetaAddr(cfg.Meta.Client.Addrs[0]),
+		service.WithMetaClient(
+			grpc.New(metadataClientOptions...),
+		),
+		service.WithMetaCacheEnabled(cfg.Meta.EnableCache),
+		service.WithMetaCacheExpireDuration(cfg.Meta.CacheExpiration),
+		service.WithMetaCacheExpiredCheckDuration(cfg.Meta.ExpiredCacheCheckDuration),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if addrs := cfg.Client.Addrs; len(addrs) == 0 {
+		return nil, errors.ErrGRPCTargetAddrNotFound
+	}
+
+	v := handler.New(
+		handler.WithClient(client.New(
+			client.WithAddr(cfg.Client.Addrs[0]),
+			client.WithClient(grpc.New(cfg.Client.Opts()...)),
+		)),
+		handler.WithMeta(metadata),
+		handler.WithErrGroup(eg),
+		handler.WithStreamConcurrency(cfg.Server.GetGRPCStreamConcurrency()),
+	)
+
+	grpcServerOptions := []server.Option{
+		server.WithGRPCRegistFunc(func(srv *grpc.Server) {
+			vald.RegisterValdServer(srv, v)
+		}),
+		server.WithPreStopFunction(func() error {
+			// TODO notify another gateway and scheduler
+			return nil
+		}),
+	}
+
+	if cfg.Observability.Enabled {
+		grpcServerOptions = append(
+			grpcServerOptions,
+			server.WithGRPCOption(
+				grpc.StatsHandler(metric.NewServerHandler()),
+			),
+		)
+	}
+
+	srv, err := starter.New(
+		starter.WithConfig(cfg.Server),
+		starter.WithREST(func(sc *config.Server) []server.Option {
+			return []server.Option{
+				server.WithHTTPHandler(
+					router.New(
+						router.WithHandler(
+							rest.New(
+								rest.WithVald(v),
+							),
+						),
+					),
+				),
+			}
+		}),
+		starter.WithGRPC(func(sc *config.Server) []server.Option {
+			return grpcServerOptions
+		}),
+		// TODO add GraphQL handler
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &run{
+		eg:            eg,
+		cfg:           cfg,
+		server:        srv,
+		observability: obs,
+		metadata:      metadata,
+	}, nil
+}
+
+func (r *run) PreStart(ctx context.Context) error {
+	if r.observability != nil {
+		return r.observability.PreStart(ctx)
+	}
+	return nil
+}
+
+func (r *run) Start(ctx context.Context) (<-chan error, error) {
+	ech := make(chan error, 6)
+	var mech, sech, oech <-chan error
+	var err error
+	if r.observability != nil {
+		oech = r.observability.Start(ctx)
+	}
+	if r.metadata != nil {
+		mech, err = r.metadata.Start(ctx)
+		if err != nil {
+			close(ech)
+			return nil, err
```

Comment:
if `oech` is not nil, do you need the function to close `oech` too?

Reply:
oech is read only channel we can't handle here.